src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -465,6 +426,60 @@ private bool TryRegister(SafeSocketHandle socket, IntPtr handle, out Interop.Err
             return error == Interop.Error.SUCCESS;
         }
 
+        // An on-stack cache of SocketAsyncEngine members needed to process epoll events.
+        // The event handling logic is delegated to a non-inlined processing method
+        // to make sure the runtime doesn't retain internal copies of SocketAsyncContext longer than necessary,
+        // so unreferenced Socket instances always get collected by the GC.
+        // See discussion: https://github.com/dotnet/runtime/issues/37064
+        private struct SocketEventHandler
+        {
+            public bool Shutdown { get; private set; }
+            public Interop.Sys.SocketEvent* Buffer { get; }
+
+            private ConcurrentDictionary<IntPtr, SocketAsyncContextWrapper> _handleToContextMap;
+            private ConcurrentQueue<SocketIOEvent> _eventQueue;
+            private IntPtr _shutdownHandle;
```

Comment:
This shouldn't need to be a field on the struct.  I expect it'd be better to change:
```C#
private static readonly IntPtr ShutdownHandle = (IntPtr)(-1);
```
to instead be:
```C#
private static IntPtr ShutdownHandle => (IntPtr)(-1);
```
and then delete `_shutdownHandle` and just use `ShutdownHandle` directly.